sys/net/gnrc/application_layer/tftp/gnrc_tftp.c

```diff
@@ -0,0 +1,868 @@
+        if(msg.type == GNRC_NETAPI_MSG_TYPE_RCV) {
+            gnrc_pktbuf_release((gnrc_pktsnip_t*) msg.content.ptr);
+        }
+    }
+
+    /* unregister our UDP listener on this thread */
+    gnrc_netreg_unregister(GNRC_NETTYPE_UDP, &entry);
+
+    return ret;
+}
+
+tftp_state _tftp_state_processes(tftp_context_t *ctxt, msg_t *m) {
+    gnrc_pktsnip_t *outbuf = gnrc_pktbuf_add(NULL, NULL, TFTP_DEFAULT_DATA_SIZE,
+                                             GNRC_NETTYPE_UNDEF);
+
+    /* check if this is an client start */
+    if (!m) {
+        return _tftp_send_start(ctxt, outbuf);
+    }
+    else if (m->type == TFTP_TIMEOUT_MSG) {
+        if (++(ctxt->retries) > GNRC_TFTP_MAX_RETRIES) {
+            /* transfer failed due to lost peer */
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* increase the timeout for congestion control */
+        ctxt->block_timeout <<= 1;
+
+        /* the send message timed out, re-sending */
+        if (ctxt->dst_port == GNRC_TFTP_DEFAULT_DST_PORT) {
+            /* we are still negotiating resent, start */
+            return _tftp_send_start(ctxt, outbuf);
+        }
+        else {
+            /* we are sending / receiving data */
+            /* if we are reading resent the ACK, if writing the DATA */
+            return _tftp_send_dack(ctxt, outbuf, (ctxt->op == TO_RRQ) ? TO_ACK : TO_DATA);
+        }
+    }
+    else if (m->type != GNRC_NETAPI_MSG_TYPE_RCV) {
+        DEBUG("tftp: unknown message");
+        gnrc_pktbuf_release(outbuf);
+        return BUSY;
+    }
+
+    gnrc_pktsnip_t *pkt = (gnrc_pktsnip_t*)(m->content.ptr);
+
+    assert(pkt->next || pkt->next->type != GNRC_NETTYPE_UDP);
+    assert(pkt->next->next || pkt->next->next->type != GNRC_NETTYPE_IPV6);
+
+    uint8_t *data = (uint8_t*)pkt->data;
+    udp_hdr_t *udp = (udp_hdr_t*)pkt->next->data;
+    ipv6_hdr_t *ip = (ipv6_hdr_t*)pkt->next->next->data;
+
+    xtimer_remove(&(ctxt->timer));
+
+    switch (_tftp_parse_type(data)) {
+    case TO_RRQ:
+    case TO_RWQ: {
+        if (byteorder_ntohs(udp->dst_port) != GNRC_TFTP_DEFAULT_DST_PORT) {
+            /* not a valid start packet */
+            gnrc_pktbuf_release(outbuf);
+
+            return FAILED;
+        }
+
+        /* reinitialize the context with the current client */
+        _tftp_init_ctxt(&(ip->src), NULL, ctxt->start_cb, ctxt->data_cb,
+                        _tftp_parse_type(data), ctxt);
+
+        /* get the context of the client */
+        ctxt->dst_port = byteorder_ntohs(udp->src_port);
+        int offset = _tftp_decode_start(ctxt, data, outbuf);
+        if (offset < 0) {
+            gnrc_pktbuf_release(outbuf);
+            return FAILED;
+        }
+
+        /* validate if the application accepts the filename and modes */
+        tftp_action_t action = ctxt->op == TO_RRQ ? TFTP_READ : TFTP_WRITE;
+        if (ctxt->start_cb(action, ctxt->file_name, ctxt->mode) != 0) {
```

Comment:
Wait, it's defined like this:
```c
typedef bool (*tftp_transfer_start_callback)(tftp_action_t action, const char *file_name, uint32_t data_len);
```

Do you say that when in server mode, the third argument is used as `tftp_action_t`? That works, because the enum is weakly typed, but that's not a good idea in general I'd say.